### PR TITLE
Introducing extensions for memcpy family of memory operations

### PR DESCRIPTION
**Pull request by pratlucas**
Following the announcement of the [2021 Architecture Extensions](https://community.arm.com/developer/ip-products/processors/b/processors-ip-blog/posts/arm-a-profile-architecture-developments-2021), I'd like to propose these aditions to the ACLE to support the new instructions that standardise the memcpy family of operations.

Here's the rationale behind the contents:

## Feature test macro

In order for users to determine the availability of the new memory operation instructions, as well of new potential intrinsics related to them, I would like to propose the addition of the following feature test macro, to be defined to 1 if the new extension is available:

```
__ARM_FEATURE_MOPS
```

## Intrinsics

Given that the purpose of the new instructions is not only optimizing but also standardising the software implementation of memcpy, memmove and memset operations, the general use of these instructions should be covered by the code generation of compilers and library implementations. Therefore, I feel there might be no motivation to introduce new intrinsics covering this scenario.

The new extension also introduces a few variations of such instructions covering the following cases:
- Privileged/unprivileged reads and writes
- Temporal/non-temporal reads and writes
- Tag setting variation of the memset operation

### Privileged/unprivileged variations

At first sight, there doesn't seem to be enough motivation to grant new intrinsics to cover those variations. Regular usage should be covered by compiler code generation and library implementation, and intrinsics for unpriviledged forms have the potential of hiding relevant implementation details from the OS.

### Temporal/non-temporal variations

Currently, in section 15.1.7 ("Memory access with non-temporal hints"), the ACLE mentions that language extensions to support cacheability hints are currently being investigated. I believe that introducing new intrinsics for non-temporal memory operations could overlap with the support under investiation, so I would suggest waiting for the investigation outcomes to provide a coherent solution.

### Tag setting memset variation

The ability to tag a memory address during a memset operation seems highly useful for MTE users - for writting a custom allocator, for example. As it's also not covered by the regular code generation and library flows, I'd like to propose the introduction of a new intrinsic to cover this scenario. I would suggest the following:

```
void* __arm_mops_memset_tag(void* tagged_address, int value, size_t size)
```
- `tagged_address`: destination address to be set, containing the allocation tag in its bits [59:56].
- `value`: fill value
- `size`: number of bytes to fill

### CODE REVIEW

File: main/acle.rst [hunk 1]
```diff
@@ -5792,3 +5811,39 @@ Instructions
...
+with memory tagging. An intrinsic is available to provide access to this
+operation. See ssec-MTE_ for more information on memory tagging.
+
+The ``<arm_acle.h>`` header should be included before using this intrinsic.
+
+Intrinsics
+==========
+
+This intrinsic is available when ``__ARM_FEATURE_MOPS`` is defined.
+::
```
Comment (fpetrogalli) [inline main/acle.rst:5841]: This should use the `.. code: c` highlighting. Also, please leave a blank line between after the text.

Comment (pratlucas) [reply main/acle.rst]: I've updated it to use the `.. code:: c` directive, but I've noticed the other snippets of the spec don't make use of it. Should this one follow the direction of the others for consistency?

Comment (fpetrogalli) [reply main/acle.rst]: Oh yes, please, do adhere to the current way code snippets are produced in this doc. Sorry about that!

Comment (pratlucas) [reply main/acle.rst:5841]: No worries, I've reverted it to match the other ones.

File: main/design_documents/__ARM_FEATURE_MOPS.rst [hunk 2]
```diff
@@ -0,0 +1,75 @@
+=============================================================
+Introducing extensions for memcpy family of memory operations
+=============================================================
+
+Following the announcement of the `2021 Architecture Extensions
+<https://community.arm.com/developer/ip-products/processors/b/processors-ip-blog/posts/arm-a-profile-architecture-developments-2021>`_,
+I'd like to propose these aditions to the ACLE to support the new instructions
```
Comment (fpetrogalli) [inline main/design_documents/__ARM_FEATURE_MOPS.rst:7]: Maybe avoid addressing this in first person, but use a generic "we"?

`We are proposing these additions to the ...`

**Review by fpetrogalli — CHANGES_REQUESTED** [review]

File: main/design_documents/__ARM_FEATURE_MOPS.rst [hunk 3]
```diff
@@ -0,0 +1,75 @@
+=============================================================
+Introducing extensions for memcpy family of memory operations
+=============================================================
+
+Following the announcement of the `2021 Architecture Extensions
+<https://community.arm.com/developer/ip-products/processors/b/processors-ip-blog/posts/arm-a-profile-architecture-developments-2021>`_,
+we are proposing these aditions to the ACLE to support the new instructions
+that standardise the memcpy family of operations.
```
Comment (fpetrogalli) [inline main/design_documents/__ARM_FEATURE_MOPS.rst:8]: Please mention `Armv8.8-A and Armv9.3-A` somewhere in this paragraph to make it clear which are the version of the architecture being mentioned here.

File: main/design_documents/__ARM_FEATURE_MOPS.rst [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+Here's the rationale behind the contents:
```
Comment (fpetrogalli) [inline main/design_documents/__ARM_FEATURE_MOPS.rst:10]: Remove

File: main/design_documents/__ARM_FEATURE_MOPS.rst [hunk 5]
```diff
@@ -0,0 +1,75 @@
+=============================================================
+Introducing extensions for memcpy family of memory operations
```
Comment (fpetrogalli) [inline main/design_documents/__ARM_FEATURE_MOPS.rst:2]:
```suggestion
Design Document for __ARM_FEATURE_MOPS
```

File: main/design_documents/__ARM_FEATURE_MOPS.rst [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+we are proposing these aditions to the ACLE to support the new instructions
+that standardise the memcpy family of operations.
+
+Here's the rationale behind the contents:
+
+Feature test macro
+##################
+
+In order for users to determine the availability of the new memory operation
+instructions, as well of new potential intrinsics related to them, we are
```
Comment (fpetrogalli) [inline main/design_documents/__ARM_FEATURE_MOPS.rst:16]:
```suggestion
In order for users to determine the availability of the memory instructions introduced in Armv8.8-A and Armv9.3-A, as well of new potential intrinsics related to them, we are
```

File: main/design_documents/__ARM_FEATURE_MOPS.rst [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+::
+
+  __ARM_FEATURE_MOPS
+
+
+Intrinsics
+##########
+
+Given that the purpose of the new instructions is not only optimizing but also
```
Comment (fpetrogalli) [inline main/design_documents/__ARM_FEATURE_MOPS.rst:28]:
```suggestion
Given that the purpose of the memory instructions in Armv8.8-A and Armv9.3-A is not only optimizing but also
```

File: main/design_documents/__ARM_FEATURE_MOPS.rst [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+standardising the software implementation of memcpy, memmove and memset
+operations, the general use of these instructions should be covered by the code
+generation of compilers and library implementations. Therefore, we feel there
+might be no motivation to introduce new intrinsics covering this scenario.
+
+The new extension also introduces a few variations of such instructions covering
```
Comment (fpetrogalli) [inline main/design_documents/__ARM_FEATURE_MOPS.rst:34]:
```suggestion
Armv8.8-A and Armv9.3-A also introduce a few variations of such instructions covering
```

Also, what do you mean with `such instructions`? memcpy, memmove and memset? If so, these are better referred as functions, not instructions?

File: main/design_documents/__ARM_FEATURE_MOPS.rst [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+the following cases:
+
+* Privileged/unprivileged reads and writes
+* Temporal/non-temporal reads and writes
+* Tag setting variation of the memset operation
```
Comment (fpetrogalli) [inline main/design_documents/__ARM_FEATURE_MOPS.rst:39]:
```suggestion
* Privileged/unprivileged reads and writes.
* Temporal/non-temporal reads and writes.
* Tag setting variation of the memset operation.
```

File: main/design_documents/__ARM_FEATURE_MOPS.rst [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+Privileged/unprivileged variations
+==================================
+
+At first sight, there doesn't seem to be enough motivation to grant new
+intrinsics to cover those variations. Regular usage should be covered by
```
Comment (fpetrogalli) [inline main/design_documents/__ARM_FEATURE_MOPS.rst:45]:
```suggestion
intrinsics to cover those variations. Regular usage is covered by
```

File: main/acle.rst [hunk 11]
```diff
@@ -2154,6 +2169,12 @@ Summary of predefined macros
    |                                     | extension           |                    |                        |
    |                                     | (32-bit-only)       |                    |                        |
    +-------------------------------------+---------------------+--------------------+------------------------+
+   | ``__ARM_FEATURE_MOPS``              | memcpy, memset, and | 1                  | ssec-MOPS_,            |
+   |                                     | memmove family of   |                    | ssec-MOPS-intrinsics_  |
+   |                                     | operations          |                    |                        |
```
Comment (fpetrogalli) [inline main/acle.rst:2174]: You can add a link in the fourth column to the design document too?

**Review by fpetrogalli — APPROVED** [review]
Thank you @pratlucas  - this LGTM. We will have one of our technical writers taking a look at it, but it would be better if before that final stage @ktkachov-arm and @DanielKristofKiss could confirm whether or not they are happy with the changes they requested.

Francesco


Thank you,

Francesco

**Comment by DanielKristofKiss** [comment]
LGTM

**Comment by pratlucas** [comment]
I've updated the contents to use the Markdown format.

The `markdown-link-check` CI run is not happy with the relative link from the design document, but it works well on Github's previews. Is there any known way to around around this issue?

File: main/design_documents/__ARM_FEATURE_MOPS.md [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+
+As the regular usage of the memory instructions is covered by compiler code
+generation and library implementation, and intrinsics for unprivileged forms
+have the potential of hiding relevant implementation details from the OS,
+there is no motivation to grant new intrinsics to cover the privileged/
+unprivileged variations.
+
+### Temporal/non-temporal variations
+
+Currently, in section [15.1.7 ("Memory access with non-temporal hints")](../acle.rst#memory-access-with-non-temporal-hints),
```
Comment (fpetrogalli) [inline main/design_documents/__ARM_FEATURE_MOPS.md:46]:
```suggestion
Currently, in section [Memory access with non-temporal hints)](../acle.md#memory-access-with-non-temporal-hints),
```

This link was detected as  broken by the CI bot

File: main/acle.md [hunk 13]
```diff
@@ -5158,4 +5213,5 @@ duplicated NEON vector `vec`.
 | `svfloat16_t svdup_neonq[_f16](float16x8_t vec)`    |
 | `svfloat32_t svdup_neonq[_f32](float32x4_t vec)`    |
 | `svfloat64_t svdup_neonq[_f64](float64x2_t vec)`    |
-| `svbfloat16_t svdup_neonq[_bf16](bfloat16x8_t vec)` |
\ No newline at end of file
+| `svbfloat16_t svdup_neonq[_bf16](bfloat16x8_t vec)` |
+
```
Comment (fpetrogalli) [inline main/acle.md:5217]: Please remove
```suggestion
```

File: main/design_documents/__ARM_FEATURE_MOPS.md [hunk 14]
```diff
@@ -0,0 +1,71 @@
+Design Document for `__ARM_FEATURE_MOPS`
+========================================
```
Comment (fpetrogalli) [inline main/design_documents/__ARM_FEATURE_MOPS.md:2]:
```suggestion
# Design Document for `__ARM_FEATURE_MOPS`
```

**Review by fpetrogalli — CHANGES_REQUESTED** [review]

File: main/acle.md [hunk 15]
```diff
@@ -5072,6 +5087,46 @@ transaction.
...
+An exception to that is the set of instructions covering the memset operation
+with memory tagging. An intrinsic is available to provide access to this
+operation. See [ssec-MTE](#memory-tagging) for more information on memory
+tagging.
+
+The `<arm_acle.h>` header should be included before using this intrinsic.
+
+This intrinsic is available when `__ARM_FEATURE_MOPS` is defined.
+
+```
```
Comment (fpetrogalli) [inline main/acle.md:5112]:
```suggestion
``` c
```

File: main/design_documents/__ARM_FEATURE_MOPS.md [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+provide a coherent solution once that investigation finishes.
+
+### Tag setting memset variation
+
+The ability to tag a memory address during a memset operation is highly useful
+for MTE users - for writing a custom allocator, for example. As it's also not
+covered by the regular code generation and library flows, we are introducing
+the following intrinsic to cover this scenario:
+
+```
```
Comment (fpetrogalli) [inline main/design_documents/__ARM_FEATURE_MOPS.md:58]:
```suggestion
``` c
```

**Review by fpetrogalli — CHANGES_REQUESTED** [review]

**Review by fpetrogalli — APPROVED** [review]
Thank you @pratlucas , LGTM.

**Pull request merged**